File: languages/typescript/02-core/maps-and-frequency-counting/example/main.ts
// Module focus: Counting repeated values and summarizing them through keyed lookups.
// Why it matters: the example makes it possible to use key-value collections to aggregate and
// retrieve data before the learner tackles the exercises.

const text = "go fast go far learn fast";
const words = text.split(/\s+/);
const frequencies = new Map<string, number>();

for (const word of words) {
    frequencies.set(word, (frequencies.get(word) ?? 0) + 1);
}

const sortedEntries = [...frequencies.entries()].sort((left, right) => {
    if (right[1] !== left[1]) {
        return right[1] - left[1];
    }
    return left[0].localeCompare(right[0]);
});

// The printed result shows whether the program can define normalization and missing-key behavior
// explicitly.
console.log(`Text: ${text}`);
for (const [word, count] of sortedEntries) {
    console.log(`${word}: ${count}`);
}

export {};

File: languages/java/02-core/maps-and-frequency-counting/example/Main.java
// Module focus: Counting repeated values and summarizing them through keyed lookups.
// Why it matters: practicing maps and frequency counting patterns makes exercises and checkpoints easier to reason about.

import java.util.Locale;
import java.util.Map;
import java.util.TreeMap;

public class Main {
    public static void main(String[] args) {
        String text = "learn java learn maps";
        Map<String, Integer> frequencies = new TreeMap<>();

        // Normalize each token before counting so equivalent words share one key.
        for (String word : text.toLowerCase(Locale.US).split("\\s+")) {
            frequencies.put(word, frequencies.getOrDefault(word, 0) + 1);
        }

        // Report sorted frequencies to keep the output stable across runs.
        System.out.println("Word frequencies:");
        for (Map.Entry<String, Integer> entry : frequencies.entrySet()) {
            System.out.println(entry.getKey() + " -> " + entry.getValue());
        }
    }
}
